File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int answer = 50;
int visited[50];// words[i]
bool checkDiff(string a, string b){
    int difCnt = 0;
    for(int i = 0 ; i < a.size(); i++){
        if(a[i] != b[i])difCnt++;
    }
    if(difCnt == 1)return true;
    return false;
}
void dfs(string begin, string target, vector<string> words, int step){
    if(answer <= step)return;
    
    if(begin == target){
        answer = min(answer, step);
        return;
    }
    
    for(int i = 0; i < words.size(); i++){
        if(checkDiff(begin, words[i]) && !visited[i]){
            visited[i] = 1;
            dfs(words[i], target, words, step + 1);
            visited[i] = 0;
        }
    }
}
int solution(string begin, string target, vector<string> words) {
    dfs(begin, target, words, 0);
    if(answer == 50)return 0;
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    static boolean[] visited;
    static int answer = Integer.MAX_VALUE;
    
    public int solution(String begin, String target, String[] words) {
        boolean flag = false;
        for(String word : words){
            if(target.equals(word))flag = true;
        }
        if(!flag)return 0;
        
        visited = new boolean[words.length];
        dfs(0, begin, target, words);
        return answer;
    }
    
    static boolean check(String a, String b){
        int cnt = 0;
        for(int i = 0; i < a.length(); i++){
            if(a.charAt(i) != b.charAt(i))cnt++;
        }
        return cnt == 1;
    }
    
    static void dfs(int cnt, String begin, String target, String[] words){
        if(answer <= cnt)return;
        if(begin.equals(target)){
            answer = Math.min(answer, cnt);
            return;
        }
        for(int i = 0; i < words.length; i++){
            if(visited[i])continue;
            if(!check(begin, words[i]))continue;
            visited[i] = true;
            dfs(cnt + 1, words[i], target, words);
            visited[i] = false;
        }
    }
}